hera_cal/io.py

```diff
@@ -864,6 +864,147 @@ def iterate_over_times(self, Nints=1, times=None):
+                   read_data=True, read_flags=False, read_nsamples=False, check=False,
+                   dtype=np.complex128, verbose=False):
+    '''A ~100x faster interface to getting data out of HERA HDF5 files. Only concatenates
+    along time axis. Assumes that if antenna_numbers header array stays the same,
+    the baseline order in the files stays the same. Puts times in ascending order,
+    but does not check that files are contiguous.
+
+    Arguments:
+        filenames: list of files to read
+        bls: list of (ant_1, ant_2, [polstr]) tuples to read out of files
+        pols: list of pol strings to read out of files
+        full_read_thresh (0.01): fractional threshold for reading whole file 
+                                 instead of baseline by baseline.
+        read_data (bool, True): read visdata
+        read_flags (bool, False): read flags
+        read_nsamples (bool, False): read nsamples
+        check (bool, False): run sanity checks to make sure files match.
+        dtype (np.complex128): numpy datatype for output complex-valued arrays
+        verbose: print some progress messages.
+
+    Returns:
+        rv: dictionary with keys 'info' (metadata), 'visdata' (dictionary of waterfalls
+            with i,j,pol keys), 'flags', and 'nsamples'. Will omit keys according to
+            read_data, read_flags, and read_nsamples.
+    '''
+    info = {}
+    times = []
+    bl2ind = {}
+    inds = {}
+    # Read file metadata to size up arrays and sort times
+    for filename in filenames:
+        if verbose:
+            print(f'Reading header of {filename}')
+        with h5py.File(filename, 'r') as f:
+            h = f['/Header']
+            if check:
+                assert int(h['Nspws'][()]) == 1  # not a hera file
+            if len(times) == 0:
+                info['freqs'] = h['freq_array'][()]
+                pol_array = h['polarization_array'][()]
+                npols = pol_array.size
+                pol_indices = {POL_NUM2STR_DICT[n]: cnt for cnt, n in enumerate(pol_array)}
+                info['pols'] = list(pol_indices.keys())
+            elif check:
+                assert int(h['Nfreqs'][()]) == info['freqs'].size
+            ntimes = int(h['Ntimes'][()])
+            times.append((h['time_array'][:ntimes], filename))
+            ant1_array = h['ant_1_array'][::ntimes]
+            ant2_array = h['ant_2_array'][::ntimes]
+            _hash = hash((ant1_array.tobytes(), ant2_array.tobytes()))
+            # map baselines to array indices for each unique antenna order
+            if _hash not in inds:
+                inds[_hash] = {(i, j): slice(n * ntimes, (n + 1) * ntimes)
+                               for n, (i, j) in enumerate(zip(ant1_array,
+                                                              ant2_array))}
+                if 'bls' not in info:
+                    info['bls'] = set(inds[_hash].keys())
```

Comment:
In pyuvdata and sometimes also in hera_cal, length-2 tuples are called `antpairs` rather than `bls` and `bls` is used for length-3 tuples that include polarizations. We're not super consistent about this though, so if you really want to call these `bls` I won't stand in your way.

Reply:
Alternatively, you can do `info['bls'] =  set(bl + (p,) for bl in bls for p in pols)` at the end of this.

Reply:
Leaving this all as stands. bls purposely not overwritten until after all files are scanned